lib/trie/trie.go

```diff
@@ -355,24 +361,40 @@ func findNextKeyChild(children []*Node, startIndex byte,
 
 // Put inserts a value into the trie at the
 // key specified in little Endian format.
-func (t *Trie) Put(keyLE, value []byte) (err error) {
+func (t *Trie) Put(keyLE, value []byte, version Version) (err error) {
 	pendingDeltas := tracking.New()
 	defer func() {
 		const success = true
 		t.handleTrackedDeltas(success, pendingDeltas)
 	}()
-	return t.insertKeyLE(keyLE, value, pendingDeltas)
+	return t.insertKeyLE(keyLE, value, pendingDeltas, version)
 }
 
 func (t *Trie) insertKeyLE(keyLE, value []byte,
-	pendingDeltas DeltaRecorder) (err error) {
+	pendingDeltas DeltaRecorder, version Version) (err error) {
 	nibblesKey := codec.KeyLEToNibbles(keyLE)
 	if value == nil {
 		// Force nil value to be inserted to []byte{} since `nil` means there
 		// is no value.
 		value = []byte{}
 	}
-	root, _, _, err := t.insert(t.root, nibblesKey, value, pendingDeltas)
+
+	isValueHashed := version.ShouldHashValue(value)
+	if isValueHashed {
+		hashedValue, err := common.Blake2bHash(value)
+		if err != nil {
+			return err
+		}
+
+		// Add original value in memory db
```

Comment:
here we don't know the database implementation exactly, so removing the `memory` makes more sense
```suggestion
		// Add original value in db and use the hash as node value
```